the-collab-lab/tcl-40-smart-shopping-list
# implements add item funtionality and list view

_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description

This code implements the user with the ability to add a new item and specify its name, date of last purchase, and frequency.  It also allows a user to see a list of all the items to the database.  We reused the function body of addTestFirebase and made it the onSubmit handler connected to our form.  We also reused the useEffect from addToFirebase to pull the list items from firebase.  We added a utility function in localStorage.js that pulls the token variable from local storage so the user can read/write to that specific list. We needed this variable name in two separate views, so we set it up this way so it can be exported into any view.

## Related Issue
closes #4 

## Acceptance Criteria
A shopping list item consists of the following data points:

 - Name of item
 - How soon are you likely to buy it again?
   - Soon (in the next 7 days)
   - Kind of soon (in the next 14 days)
   - Not soon (in the next 30 days)
   - Last purchased date

 - User is presented with a form that allows them to enter the name of the item and select how soon they anticipate needing to buy it again (Soon, Kind of Soon, Not Soon)
 - When the user submits the form, the item is saved to the database, associated with the user’s token
Along with the item name, an integer corresponding to the estimated number of days until next purchase is saved: 7 for “Soon”, 14 for “Kind of Soon”, and 30 for “Not Soon”
- The last purchased date should be set to null initially (i.e. you can create an item without purchasing it)
- Item names should be displayed as a simple, unordered list on the list view after they’re added
Notes:
- The add item form has the potential to be a lot of scope. The wireframe https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/add-item.png has several notes related to accessibility and UX.
- Eventually, the list will look like the following wireframe: https://github.com/the-collab-lab/smart-shopping-list/blob/main/_resources/wireframes/list-populated.png
It's good for your team to look at the wireframe now, but be sure not to try to implement all of the visible functionality at once. There will be future stories to add all of those features.
- For this story, it's sufficient just to get a simple list of the items displaying.

## Type of Changes


|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before

### After


![Screenshot (114)](https://user-images.githubusercontent.com/83678228/163020030-3d8c91bf-c29f-4d4a-8c13-92a3117caecc.png)
![Screenshot (115)](https://user-images.githubusercontent.com/83678228/163020049-1d815cd3-4b6d-4b8d-9759-16e026ab2d24.png)

## Testing Steps / QA Criteria
In the App
 - pull down the changes from our branch
   -  ` git pull`
- checkout our branch
   -  `git checkout nj-ks-addItem`
- run the server
   -  `npm start`
- navigate to the addItem view in the app
- add an item in the input field
- select a frequency
- date should be left blank (but if you select you should see it the database)
-  navigate to the list view in the app
   - you should see an unordered list of items that have been added to the list.
   -  the most recently added item will appear at the bottom
- comparing this list view with what's in firebase, you should see the same items that are associated with the list named "KN"
   
In Firebase
- go firebase to see our collection by clicking the following link:
- [firebase](https://console.firebase.google.com/u/0/project/tcl-40-smart-shopping-list/firestore/data/~2Flist~2F9b5VKKLyJPyEmiX9PHRr)
-  click on "KN" 
  - (we created this dummy variable "KN" to simulate the user token in local storage.  When we merge our branches together the name of the list in firebase will change to whatever token is generated from the other teams branch.)
  - select any document to see its properties
  -  you should see a list item with the following properties: name, frequency, and date
- to update an item in firebase
  - click on the 3 vertical dots icon in the right most column, mouse over the property you want to edit and click the pencil icon.  
  - update the text and click the save button
  - go back to the app and it should have updated automatically.

![Screenshot (117)](https://user-images.githubusercontent.com/83678228/163274354-9ebc6a40-5923-49ec-8732-4e0b54956b44.png)
